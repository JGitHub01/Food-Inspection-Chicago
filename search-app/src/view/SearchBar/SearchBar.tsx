import { useRef, useEffect } from 'react';
import { TextField, Button } from '@material-ui/core';
import './styles/search-bar.scss';

interface ISearchBarProps {
  keyword?: string
  onKeywordChange?: (keyword: string) => void
  onSearch?: (keyword: string) => void
}

export default function SearchBar(props: ISearchBarProps) {
  const { keyword = "", onKeywordChange, onSearch } = props;
  const buttonRef = useRef<HTMLButtonElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  useEffect(() => {
    const onButtonClick = () => {
      onSearch && onSearch(keyword);
    };
    const onKeypress = (e: KeyboardEvent) => {
      if (e.key === 'Enter') {
        onSearch && onSearch(keyword);
      }
    }
    buttonRef.current?.addEventListener('click', onButtonClick);
    inputRef.current?.addEventListener('keypress', onKeypress);

    return (() => {
      buttonRef.current?.removeEventListener('click', onButtonClick);
      inputRef.current?.removeEventListener('keypress', onKeypress);
    });
  }, [keyword]);

  return (
    <div className='search-bar' data-testid='search-bar'>
      <TextField
        className='search-bar__text'
        label='Type a keyword'
        variant='filled'
        inputRef={inputRef}
        inputProps={{'data-testid': 'search-field'}}
        onChange={(e) => {
          onKeywordChange && onKeywordChange(e.target.value);
        }}
      >
        {keyword}
      </TextField>
      <div className='search-bar__button'>
        <Button
          variant='contained'
          color='primary'
          ref={buttonRef}
          data-testid='search-button'
        >
          Search
        </Button>
      </div>
    </div>
  );
}
