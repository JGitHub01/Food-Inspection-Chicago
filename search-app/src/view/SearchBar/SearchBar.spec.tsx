import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SearchBar from './SearchBar';

describe('SearchBar', () => {
  it('Should render the default snapshort view', () => {
    const { getByTestId } = render(<SearchBar keyword='fried' />);
    const searchBar = getByTestId('search-bar');
    expect(searchBar).toMatchSnapshot();
  });

  it('Should call onKeywordChange with the entered text when a keyword is entered', () => {
    const callback = jest.fn();
    const { getByTestId } = render(<SearchBar onKeywordChange={callback} />);
    const searchField = getByTestId('search-field');
    userEvent.type(searchField, 'fried');
    expect(callback).toBeCalledWith('fried');
  });

  it('Should call onSearch with the keyword when hit enter', () => {
    const callback = jest.fn();
    const { getByTestId } = render(<SearchBar keyword='fried' onSearch={callback} />);
    const searchField = getByTestId('search-field');
    userEvent.type(searchField, '{enter}');
    expect(callback).toBeCalledWith('fried');
  });

  it('Should call onSearch with keyword when search button is clicked', () => {
    const callback = jest.fn();
    const { getByTestId } = render(<SearchBar keyword='fried' onSearch={callback} />);
    const searchButton = getByTestId('search-button');
    userEvent.click(searchButton);
    expect(callback).toBeCalledWith('fried');
  });
});
