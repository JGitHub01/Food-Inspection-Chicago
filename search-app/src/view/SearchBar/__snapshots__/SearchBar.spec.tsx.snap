// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Should render the default snapshort view 1`] = `
<div
  class="search-bar"
  data-testid="search-bar"
>
  <div
    class="MuiFormControl-root MuiTextField-root search-bar__text"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
      data-shrink="false"
    >
      Type a keyword
    </label>
    <div
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiFilledInput-input"
        data-testid="search-field"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="search-bar__button"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      data-testid="search-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Search
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
